### PR TITLE
invitelink more annotations

### PR DESCRIPTION
**Pull request by heronhaye**

### CODE REVIEW

File: go/client/team_cli_rendering.go [hunk 1]
```diff
@@ -187,9 +203,6 @@ func (c *teamMembersRenderer) outputTeams(list keybase1.AnnotatedTeamList, showA
 			fmt.Fprintf(c.tabw, "%s\t%s\t%d\n", t.FqName, role, t.MemberCount)
 		}
 	}
-	if showAll {
-		c.outputInvites(list.AnnotatedActiveInvites)
```
Comment (heronhaye) [inline go/client/team_cli_rendering.go:191]: behavior change

File: go/teams/list.go [hunk 2]
```diff
@@ -493,38 +477,62 @@ func ComputeInvitelinkDisplayName(mctx libkb.MetaContext, team *Team, invite key
...
 		ret = append(ret, keybase1.AnnotatedTeamUsedInviteLogPoint{TeamUsedInviteLogPoint: point, Username: username.String()})
 	}
 	return ret
 }
 
-type AnnotatedInviteMap map[keybase1.TeamInviteID]keybase1.AnnotatedTeamInvite
+// GetAnnotatedInvitesAndMembersForUI returns annotated invites and members lists from the sigchain,
+// except that Keybase-type invites are treated as team members. This is for the purpose of
+// collecting them together in the UI.
+func GetAnnotatedInvitesAndMembersForUI(mctx libkb.MetaContext, team *Team,
```
Comment (heronhaye) [inline go/teams/list.go:491]: large changes here. Note it's returning every invite now, even cancelled/completed/obsolete ones.

File: go/teams/list.go [hunk 3]
```diff
@@ -595,31 +600,16 @@ func AnnotateInvitesNoPUKless(mctx libkb.MetaContext, team *Team,
...
-					members.Admins = addKeybaseInviteToRes(mctx.Ctx(), details, members.Admins)
-				case keybase1.TeamRole_WRITER:
-					members.Writers = addKeybaseInviteToRes(mctx.Ctx(), details, members.Writers)
-				case keybase1.TeamRole_READER:
-					members.Readers = addKeybaseInviteToRes(mctx.Ctx(), details, members.Readers)
-				}
-			}
-			// Continue to skip adding this invite to annotatedInvites
-			continue
+			displayName = keybase1.TeamInviteDisplayName(pkg.NormalizedUsername.String())
```
Comment (heronhaye) [inline go/teams/list.go:606]: we're adding to invites map here; i.e., this function is not "Lying" about kb invites

File: go/teams/service_helper.go [hunk 4]
```diff
@@ -31,153 +30,79 @@ func LoadTeamPlusApplicationKeys(ctx context.Context, g *libkb.GlobalContext, id
...
-
-	name := t.Data.Name.String()
+	teamNameStr := t.Data.Name.String()
 	myRole, err := t.myRole(ctx)
-	var isAdmin bool
 	if err != nil {
-		g.Log.CDebugf(ctx, "Error getting role; skipping getting requests: %v", err)
-	} else {
-		isAdmin = myRole.IsOrAbove(keybase1.TeamRole_ADMIN)
+		return res, err
```
Comment (heronhaye) [inline go/teams/service_helper.go:66]: i'm making this error out but i don't see why it should ever error

File: go/teams/service_helper.go [hunk 5]
```diff
@@ -31,153 +30,79 @@ func LoadTeamPlusApplicationKeys(ctx context.Context, g *libkb.GlobalContext, id
...
-		tarsDisabled, err = GetTarsDisabled(ctx, g, id)
+		maybeJoinRequests = &joinRequests
+		tarsDisabled, err := GetTarsDisabled(ctx, g, teamID)
 		if err != nil {
 			return res, err
 		}
+		maybeTARsDisabled = &tarsDisabled
 	}
 
-	showcase, err := GetTeamShowcase(ctx, g, id)
```
Comment (heronhaye) [inline go/teams/service_helper.go:104]: no more showcase here, need to call another rpc (we can merge later)

File: go/service/teams.go [hunk 6]
```diff
@@ -231,51 +232,22 @@ func (h *TeamsHandler) GetAnnotatedTeam(ctx context.Context, arg keybase1.TeamID
 	return teams.GetAnnotatedTeam(ctx, h.G().ExternalG(), arg)
 }
 
-func (h *TeamsHandler) teamGet(ctx context.Context, details keybase1.TeamDetails, teamDescriptor string) (keybase1.TeamDetails, error) {
-	if details.Settings.Open {
-		h.G().Log.CDebugf(ctx, "TeamGet: %q is an open team, filtering reset writers and readers", teamDescriptor)
-		details.Members.Writers = keybase1.FilterInactiveMembers(details.Members.Writers)
```
Comment (heronhaye) [inline go/service/teams.go:237]: This now happens in GetAnnotatedTeam

File: shared/teams/team/invites/invite-item.tsx [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+}) => {
+  const dispatch = Container.useDispatch()
+  const yourUsername = Container.useSelector(s => s.config.username)
+  const [waitingForExpire, setWaitingForExpire] = React.useState(false)
+  const onExpire = () => {
+    dispatch(TeamsGen.createRemovePendingInvite({inviteID: inviteLink.id, teamID}))
+
+    // wait until reload happens due to MetadataUpdate notification; otherwise it flashes
+    // active in between the rpc finish and the team reload
+    setWaitingForExpire(true)
```
Comment (heronhaye) [inline shared/teams/team/invites/invite-item.tsx:27]: This is a new change I made.

Comment (heronhaye) [reply shared/teams/team/invites/invite-item.tsx:27]: The rest of this function is pretty much the same except for some switching on the mode of where it's used, and swapping out expiration formatting with validityDescription

File: protocol/avdl/keybase1/teams.avdl [hunk 8]
```diff
@@ -643,19 +644,41 @@ protocol teams {
...
-    union { null, TeamMemberStatus } status;
+    // isValid is whether an invite can still be completed.
+    // For single use invites, it means they are active and not completed or cancelled yet.
+    // For invitelinks, it means that they are active and (if an expiration time has been set) it
+    // has not expired, according to the client's local clock, and (if a max uses has been set) the
+    // maximum number of uses has not been reached.
+    boolean isValid;
+    // validityDescription is a description like "Expires after 12 days or 3 uses",
+    // "Expired after 12 uses", "Expired 100 years ago", "Does not expire".
+    string validityDescription;
```
Comment (buoyad) [inline protocol/avdl/keybase1/teams.avdl:662]: Hesitant about this change vs. creating this text in the gui. If the invite is expiring in on the order of hours you'd likely see a fairly outdated time here. What do you think?

Comment (heronhaye) [reply protocol/avdl/keybase1/teams.avdl:662]: Right, that's why Y2K-1722 is going to send a notification on expiry or reload notifications if the link is expiring soon. Like reload every 30 minutes if its expiring in a day, every minute if its less than 2 hours, etc.

Comment (heronhaye) [reply protocol/avdl/keybase1/teams.avdl:662]: for < 2 minutes we can either just do "Expiring soon" or reload every 3-4 seconds. I don't think the exact time is as important here compared to ephemeral messages

Comment (buoyad) [reply protocol/avdl/keybase1/teams.avdl:662]: I agree the it's not that important. The notification loop seems relatively tricky for something that might be solved by putting a timestamp field here, though.

File: shared/constants/teams.tsx [hunk 9]
```diff
@@ -67,39 +67,26 @@ export const initialMemberInfo = Object.freeze<Types.MemberInfo>({
...
-      infos.push([
+  members.forEach(({fullName, joinTime, status, username, role}) => {
+    const maybeRole = teamRoleByEnum[role]
+    infos.push([
+      username,
+      {
+        fullName,
+        joinTime: joinTime || undefined,
+        status: rpcMemberStatusToStatus[status],
+        type: !maybeRole || maybeRole === 'none' ? 'reader' : maybeRole,
```
Comment (buoyad) [inline shared/constants/teams.tsx:81]: can `maybeRole === 'none'` happen? are we risking showing wrong if we change it to reader?

File: shared/constants/types/teams.tsx [hunk 10]
```diff
@@ -62,7 +62,7 @@ export type MemberInfoWithLastActivity = MemberInfo & {
   lastActivity?: number
 }
 
-export type InviteInfo = {
+export type ValidInviteInfo = {
```
Comment (buoyad) [inline shared/constants/types/teams.tsx:65]: Prefer not to change this

File: shared/constants/teams.tsx [hunk 11]
```diff
@@ -991,3 +979,23 @@ export const maybeGetSparseMemberInfo = (state: TypedState, teamID, username) =>
   }
   return state.teams.treeLoaderTeamIDToSparseMemberInfos.get(teamID)?.get(username)
 }
+
+export const countValidInviteLinks = (inviteLinks: Array<Types.InviteLink>): Number => {
```
Comment (buoyad) [inline shared/constants/teams.tsx:983]: `reduce` is appropriate here

File: shared/constants/teams.tsx [hunk 12]
```diff
@@ -991,3 +979,23 @@ export const maybeGetSparseMemberInfo = (state: TypedState, teamID, username) =>
...
+
+export const maybeGetMostRecentValidInviteLink = (
+  inviteLinks: Array<Types.InviteLink>
+): Types.InviteLink | undefined => {
+  const validInviteLinks = inviteLinks.filter(inviteLink => inviteLink.isValid)
+  if (!validInviteLinks) {
+    return undefined
+  }
+  return validInviteLinks[0]
+}
```
Comment (buoyad) [inline shared/constants/teams.tsx:1001]:
```suggestion
) => inviteLinks.find(inviteLink => inviteLink.isValid)
```

File: shared/constants/types/teams.tsx [hunk 13]
```diff
@@ -116,8 +115,10 @@ export type InviteLink = {
 export type TeamDetails = {
   members: Map<string, MemberInfo>
   settings: TeamSettings2
-  invites: Set<InviteInfo>
-  inviteLinks: Set<InviteLink>
+  // Legacy invites that are guaranteed to be active
+  validInvites: Set<ValidInviteInfo>
```
Comment (buoyad) [inline shared/constants/types/teams.tsx:119]: Prefer not to change this too. Is there a reason not to change this to array also?

File: shared/teams/team/invites/generate-link.tsx [hunk 14]
```diff
@@ -225,9 +171,9 @@ const GenerateLinkModal = (props: Props) => {
             Here is your link. Share it cautiously as anyone who has it can join the team.
           </Kb.Text>
 
-          <InviteItem duration={inviteDuration} inviteLink={inviteLink} teamID={teamID} />
+          <InviteItem inviteLink={inviteLink} teamID={teamID} mode={'generate'} />
```
Comment (buoyad) [inline shared/teams/team/invites/generate-link.tsx:174]:
```suggestion
          <InviteItem inviteLink={inviteLink} teamID={teamID} mode="generate" />
```

File: shared/teams/team/invites/invite-history.tsx [hunk 15]
```diff
@@ -101,17 +99,17 @@ const InviteHistory = (props: Props) => {
...
+        selectedTab={showingValid ? validTitle : invalidTitle}
         style={styles.tabs}
       />
       {emptyOrLoading ?? (
         <Kb.BoxGrow>
           <Kb.SectionList
             sections={sections}
             keyExtractor={item => item.id}
-            renderItem={({item}) => <InviteItem inviteLink={item} teamID={teamID} />}
+            renderItem={({item}) => <InviteItem inviteLink={item} teamID={teamID} mode={'history'} />}
```
Comment (buoyad) [inline shared/teams/team/invites/invite-history.tsx:112]:
```suggestion
            renderItem={({item}) => <InviteItem inviteLink={item} teamID={teamID} mode="history" />}
```

File: shared/teams/team/invites/invite-item.tsx [hunk 16]
```diff
@@ -0,0 +1,117 @@
+import * as React from 'react'
+import * as Kb from '../../../common-adapters'
+import * as Styles from '../../../styles'
+import * as Container from '../../../util/container'
+import * as Types from '../../../constants/types/teams'
+import * as TeamsGen from '../../../actions/teams-gen'
+
+export type Mode = 'header' | 'settings' | 'history' | 'generate'
```
Comment (buoyad) [inline shared/teams/team/invites/invite-item.tsx:8]: I think we should ditch this component knowing about where it's used in favor of this taking a `style` and prop to hide the "created by" line

File: shared/teams/team/invites/invite-item.tsx [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+    borderRadius: Styles.borderRadius,
+    borderStyle: 'solid',
+    borderWidth: 1,
+    padding: Styles.globalMargins.tiny,
+  },
+  wideMargins: {
+    marginLeft: Styles.globalMargins.small,
+    marginRight: Styles.globalMargins.small,
+    marginTop: Styles.globalMargins.small,
+  },
```
Comment (buoyad) [inline shared/teams/team/invites/invite-item.tsx:116]: already covered by last comment but this kind of style is usually better to define in the parent

File: shared/teams/team/settings-tab/invite-links.tsx [hunk 18]
```diff
@@ -26,41 +25,19 @@ const InviteLinks = (props: Props) => {
...
-            Invites as {mostRecentInviteLink.role} · Expires after{' '}
-            {formatExpirationTimeForInviteLink(mostRecentInviteLink.expirationTime)}
-          </Kb.Text>
-        </Kb.Box2>
+        <InviteItem inviteLink={mostRecentInviteLink} teamID={teamID} mode={'settings'} />
       )}
       <Kb.Box2 direction={Styles.isMobile ? 'vertical' : 'horizontal'} gap="tiny" alignSelf="flex-start">
         <Kb.Button small={true} mode="Secondary" label="Generate invite link" onClick={onGenLink} />
-        <Kb.Button small={true} mode="Secondary" label="View link history" onClick={onViewHistory} />
+        <Kb.Button small={true} mode="Secondary" label={`Manage invite links`} onClick={onViewHistory} />
```
Comment (buoyad) [inline shared/teams/team/settings-tab/invite-links.tsx:37]:
```suggestion
        <Kb.Button small={true} mode="Secondary" label="Manage invite links" onClick={onViewHistory} />
```

File: shared/teams/team/invites/invite-item.tsx [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+          </Kb.Box2>
+        )}
+      </Kb.Box2>
+    </Kb.Box2>
+  )
+}
+
+const styles = Styles.styleSheetCreate(() => ({
+  disabledText: {opacity: 0.4},
+  inviteContainer: {
```
Comment (buoyad) [inline shared/teams/team/invites/invite-item.tsx:105]: I think this one makes most sense as the default style for this, the header's style can override this away

**Comment by heronhaye** [comment]
Thanks @buoyad, made changes!

File: go/teams/service_helper.go [hunk 20]
```diff
@@ -31,268 +30,176 @@ func LoadTeamPlusApplicationKeys(ctx context.Context, g *libkb.GlobalContext, id
...
+		} else {
+			g.Log.CDebugf(ctx, "membersHideInactiveDuplicates filtered out row: %v %v", member.Uv, member.Status)
 		}
-		*rows = filtered
 	}
+	return ret
 }
 
-func MembersDetails(ctx context.Context, g *libkb.GlobalContext, t *Team) (ret keybase1.TeamMembersDetails, err error) {
+func MembersDetails(ctx context.Context, g *libkb.GlobalContext, t *Team) (ret []keybase1.TeamMemberDetails, err error) {
```
Comment (zapu) [inline go/teams/service_helper.go:160]: might as well change it to take mctx while we are there

File: go/teams/service_helper.go [hunk 21]
```diff
@@ -319,11 +226,16 @@ func userVersionsToDetails(mctx libkb.MetaContext, uvs []keybase1.UserVersion, t
...
+		}
 		ret[i] = keybase1.TeamMemberDetails{
 			Uv:       uvs[i],
 			Username: pkg.NormalizedUsername.String(),
 			FullName: fullName,
 			Status:   status,
+			Role:     role,
 		}
 		if status == keybase1.TeamMemberStatus_ACTIVE && t != nil {
 			joinTime, err := t.UserLastJoinTime(uv)
```
Comment (zapu) [inline go/teams/service_helper.go:241]: so while we are here, can we figure out duplication of 
```
func (t TeamSigChainState) MemberCtime(user keybase1.UserVersion) *keybase1.Time
```
```
func (s TeamSigChainState) GetUserLastJoinTime(user UserVersion) (time Time, err error)
```
?

IIRC it's has the same function, so we can just do with one.

File: go/kbtime/humanize.go [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+			lbl = " " + lbl
+		}
+		pl := ""
+		if yearDiff > 1 {
+			pl = "s"
+		}
+		return fmt.Sprintf("%d year%s%s", yearDiff, pl, lbl)
+	}
+
+	return strings.TrimSuffix(humanize.RelTime(a, b, albl, blbl), " ")
```
[the remark this answers is not in the document]
Comment (heronhaye) [reply go/kbtime/humanize.go:43]: this library is not great, it'll add a trailing space even if you dont put anything in the albl blbl

File: go/teams/list.go [hunk 23]
```diff
@@ -493,38 +477,86 @@ func ComputeInvitelinkDisplayName(mctx libkb.MetaContext, team *Team, invite key
...
+		inviteC, err := annotatedInvite.InviteMetadata.Invite.Type.C()
+		if err != nil {
+			return nil, nil, err
+		}
+		extC, err := annotatedInvite.InviteExt.C()
+		if err != nil {
+			return nil, nil, err
+		}
+		if inviteC != extC {
+			return nil, nil, fmt.Errorf("got invite category %v from invite but %v from inviteExt", inviteC, extC)
```
[the remark this answers is not in the document]
Comment (heronhaye) [reply go/teams/list.go:513]: Not sure what you're suggesting exactly, but either sound fine to me (continue or error).

File: go/teams/service_helper.go (hunk 21 above)
Comment (heronhaye) [reply go/teams/service_helper.go:241]: Ticket

Comment (heronhaye) [reply go/teams/service_helper.go:241]: (it seems not immediately obvious)

File: go/teams/service_helper.go (hunk 20 above)
Comment (heronhaye) [reply go/teams/service_helper.go:160]: Gave this a shot but its being called from chat where i'd be uncomfortable with making a libkb.NewMetaContext

**Pull request merged**